<a href="https://colab.research.google.com/github/tejokumart/advanced-python-projects-build-ai-applications-4465602-1/blob/main/Readtextfile.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Perform text similarity analysis on a user-provided text file to find the most similar strings to a given input string using TF-IDF.

## Load Text Data

### Subtask:
Read the text strings line by line from a user-specified file into a list. The user will be prompted to provide the filename.


**Reasoning**:
To prompt the user for a filename and then read the text file line by line into a list, handling potential file not found errors, I will use a Python code block.



# Task
Read the text data from the `data.txt` file, which was specified in the previous step, line by line into a list called `text_data`.

## Upload and Read Text Data

### Subtask:
Allow the user to upload a text file. The content of the uploaded file will then be read line by line into a list called `text_data`.


**Reasoning**:
To fulfill the subtask requirements, I will generate a Python code block that attempts to open the file specified by `file_name`. It will handle `FileNotFoundError`, read lines into the `text_data` list while stripping whitespace, and finally print the length of the list for confirmation.



In [66]:
try:
    with open(file_name, 'r') as file:
        text_data = [line.strip() for line in file if line.strip()]
    print(f"Successfully loaded {len(text_data)} lines from '{file_name}'.")
    if text_data:
        print("First 5 lines:")
        for i, line in enumerate(text_data[:5]):
            print(f"  {i+1}: {line}")
except FileNotFoundError:
    print(f"Error: The file '{file_name}' was not found. Please make sure the file is in the correct directory.")
except Exception as e:
    print(f"An unexpected error occurred: {e}")

Successfully loaded 1667 lines from 'data.txt'.
First 5 lines:
  1: Vendor circuit Id
  2: 81-I0V5CT-XCON
  3: 81-2JGW7GJ-XCON
  4: 81-2JGW7GJ-XCON
  5: 81-I0V5CT-XCON


## Create and Train TF-IDF Model

### Subtask:
Initialize a TfidfVectorizer and fit it to the loaded text data. This step builds the vocabulary and learns the term frequencies and inverse document frequencies, which are essential for calculating text similarity.


**Reasoning**:
To initialize the TfidfVectorizer, fit it to the text data, and transform the data into TF-IDF vectors, I need to import the class from scikit-learn and then apply the fit_transform method to the `text_data`.



In [71]:
from sklearn.feature_extraction.text import TfidfVectorizer
import re # Import the regular expression module

def custom_preprocessor(text):
    text = text.lower()
    # Remove 'var_' prefix if present, only if it's at the beginning
    if text.startswith('var_'):
        text = text[4:]

    # Replace all non-alphanumeric characters (except spaces for now) with a single space
    # This handles _, /, ., -, etc.
    text = re.sub(r'[^a-z0-9]+', ' ', text)

    # Remove leading/trailing spaces and multiple spaces
    text = re.sub(r'\s+', ' ', text).strip()

    return text

# Initialize TfidfVectorizer with custom preprocessor
tfidf_vectorizer = TfidfVectorizer(preprocessor=custom_preprocessor)

# Fit and transform the text data
tfidf_matrix = tfidf_vectorizer.fit_transform(text_data)

print("TF-IDF Vectorizer initialized and fitted with custom preprocessor.")
print(f"Shape of TF-IDF matrix: {tfidf_matrix.shape}")

TF-IDF Vectorizer initialized and fitted with custom preprocessor.
Shape of TF-IDF matrix: (1667, 1763)


## Get User Input and Vectorize

### Subtask:
Prompt the user to enter the input string they want to analyze for similarity. This input string will then be transformed into its TF-IDF vector representation using the previously trained TfidfVectorizer.


**Reasoning**:
To prompt the user for an input string and transform it into a TF-IDF vector, I will use a Python code block. This will involve using the `input()` function to get the string and the previously fitted `tfidf_vectorizer` to convert the string into its TF-IDF representation.



In [79]:
user_input_string = input("Please enter the string you want to analyze for similarity: ")

# Transform the user input string into its TF-IDF vector representation
user_input_tfidf = tfidf_vectorizer.transform([user_input_string])

print(f"User input: '{user_input_string}'")
print(f"Shape of user input TF-IDF vector: {user_input_tfidf.shape}")

Please enter the string you want to analyze for similarity: PacNet ID: AII-0018-1/FE2-C463
User input: 'PacNet ID: AII-0018-1/FE2-C463'
Shape of user input TF-IDF vector: (1, 1763)


## Calculate Similarity

### Subtask:
Calculate the cosine similarity between the user's input TF-IDF vector and all the TF-IDF vectors of the `text_data`.

#### Instructions
1. Import the `cosine_similarity` function from `sklearn.metrics.pairwise`.
2. Calculate the cosine similarity between `user_input_tfidf` and `tfidf_matrix`.
3. Store the similarity scores in a variable, for example, `cosine_similarities`.

**Reasoning**:
To calculate the cosine similarity, I need to import the `cosine_similarity` function and then apply it to the `user_input_tfidf` and `tfidf_matrix`.



In [80]:
from sklearn.metrics.pairwise import cosine_similarity

# Calculate cosine similarity between the user input TF-IDF vector and all text data TF-IDF vectors
cosine_similarities = cosine_similarity(user_input_tfidf, tfidf_matrix)

print(f"Shape of cosine similarities: {cosine_similarities.shape}")

Shape of cosine similarities: (1, 1667)


## Display Most Similar Strings

### Subtask:
Retrieve and display the top N most similar strings from the original `text_data` based on the calculated cosine similarities. The user should be able to specify the number of top similar strings to display.

#### Instructions
1. Prompt the user to enter the number of top similar strings they want to see.
2. Get the indices of the top N most similar strings.
3. Print the top N similar strings along with their similarity scores.

**Reasoning**:
To display the top N most similar strings, I need to prompt the user for the number of results, sort the cosine similarities in descending order, and then retrieve and print the corresponding strings from `text_data` along with their scores.



In [81]:
import numpy as np

try:
    num_similar = int(input("Enter the number of top similar strings to display: "))
except ValueError:
    print("Invalid input. Please enter an integer.")
    num_similar = 5 # Default to 5 if input is invalid

# Flatten the cosine similarities array to easily sort
similarity_scores = cosine_similarities.flatten()

# Get the indices of the top N most similar strings (in descending order)
top_n_indices = similarity_scores.argsort()[-num_similar:][::-1]

print(f"\nTop {num_similar} most similar strings to '{user_input_string}':")
for i, index in enumerate(top_n_indices):
    score = similarity_scores[index]
    original_string = text_data[index]
    print(f"  {i+1}. Score: {score:.4f}, String: '{original_string}'")

Enter the number of top similar strings to display: 3

Top 3 most similar strings to 'PacNet ID: AII-0018-1/FE2-C463':
  1. Score: 0.4191, String: 'Vendor circuit Id'
  2. Score: 0.3707, String: 'AII-0020-YB4-COLO'
  3. Score: 0.3707, String: 'AII-0020-AB-COLO'


### Investigating the input string's presence in `text_data`

Let's check if the string `'105T1ZFFTSMARSUH10FTSMARSUK06'` actually exists in the `text_data` list, and how its components might be tokenized by the `TfidfVectorizer`.

In [78]:
is_present = user_input_string in text_data
print(f"Is '{user_input_string}' directly present in text_data? {is_present}")

# Let's also check if any of the tokens from the input string are in the TF-IDF vocabulary
input_tokens = tfidf_vectorizer.build_tokenizer()(user_input_string)
print(f"\nTokens generated from input string: {input_tokens}")

vocabulary_words = tfidf_vectorizer.get_feature_names_out()

found_tokens = [token for token in input_tokens if token in vocabulary_words]
print(f"Tokens from input string found in TF-IDF vocabulary: {found_tokens}")

if not found_tokens:
    print("\nNone of the tokens from the input string were found in the TF-IDF model's vocabulary.")
    print("This is why the similarity score was 0.0000, as there are no common terms between the input and the documents in the corpus according to the TF-IDF model.")
else:
    print("\nSome tokens from the input string were found in the TF-IDF model's vocabulary, but their distribution might not lead to high similarity scores.")


Is 'PF-CC-DAL-SFO-1805325-PF / 1808404' directly present in text_data? False

Tokens generated from input string: ['PF', 'CC', 'DAL', 'SFO', '1805325', 'PF', '1808404']
Tokens from input string found in TF-IDF vocabulary: ['1805325', '1808404']

Some tokens from the input string were found in the TF-IDF model's vocabulary, but their distribution might not lead to high similarity scores.


## Final Task

### Subtask:
Provide a summary of the text similarity analysis, including the most similar string found and any other relevant insights.


## Summary:

### Data Analysis Key Findings

*   A text file named `data.txt` containing 1667 lines was successfully loaded into a list called `text_data`. Each line was stripped of leading/trailing whitespace.
*   A `TfidfVectorizer` was initialized and fitted to the `text_data`, resulting in a TF-IDF matrix with a shape of (1667, 1859). This indicates 1667 documents and 1859 unique terms were identified.
*   Upon user input of the string `'AH827369-MPLS'`, it was transformed into a TF-IDF vector of shape (1, 1859).
*   Cosine similarity was calculated between the user's input and all 1667 documents in `text_data`.
*   The most similar string found for the input `'AH827369-MPLS'` was an exact match: `'AH827369-MPLS'`, with a perfect similarity score of 1.0000.

### Insights or Next Steps

*   The analysis successfully demonstrates the use of TF-IDF and cosine similarity for finding exact matches within a dataset. To make this more robust, consider analyzing cases where the input string isn't an exact match in the dataset to evaluate the model's ability to find approximate or semantically similar entries.
*   Explore incorporating fuzzy matching techniques (e.g., Levenshtein distance) in conjunction with TF-IDF for scenarios where partial matches or typos are expected, which could enhance the discovery of similar strings even if they are not identical.
